yamcs-core/src/main/java/org/yamcs/security/SdlsSecurityAssociation.java

```diff
@@ -0,0 +1,356 @@
+                                                  byte[] partialAuthMask) {
+        // Read security header
+        // first two bytes are SPI
+        int secHeaderStart = dataStart - getHeaderSize();
+        short receivedSpi = (short) ByteArrayUtils.decodeUnsignedShort(transferFrame, secHeaderStart);
+
+        // Check that the received SPI is the SPI for this SA
+        if (receivedSpi != spi) {
+            log.error("Expected SPI {}, received SPI {}", spi, receivedSpi);
+            return VerificationStatusCode.InvalidSPI;
+        }
+
+        // Next bytes of the header are IV
+        byte[] receivedIv = new byte[GCM_IV_LEN_BYTES];
+        System.arraycopy(transferFrame, secHeaderStart + 2, receivedIv, 0, GCM_IV_LEN_BYTES);
+
+        // create data to authenticate by masking frame headers with authMask
+        byte[] authMask = completeAuthMask(partialAuthMask);
+        byte[] aad = new byte[authMask.length];
+        for (int i = 0; i < authMask.length; ++i) {
+            aad[i] = (byte) (transferFrame[frameStart + i] & authMask[i]);
+        }
+
+        // Initialize the cipher
+        final Cipher cipher;
+        try {
+            cipher = Cipher.getInstance(cipherName);
+        } catch (NoSuchAlgorithmException | NoSuchPaddingException e) {
+            return VerificationStatusCode.NoSuchCipher;
+        }
+
+        // Configure the cipher with our parameters, the received IV, and our key
+        AlgorithmParameterSpec gcmIv = new GCMParameterSpec(GCM_TAG_LEN_BITS, receivedIv, 0, GCM_IV_LEN_BYTES);
+        try {
+            cipher.init(Cipher.DECRYPT_MODE, secretKey, gcmIv);
+        } catch (InvalidKeyException e) {
+            return VerificationStatusCode.InvalidCipherKey;
+        } catch (InvalidAlgorithmParameterException e) {
+            return VerificationStatusCode.InvalidCipherParam;
+        }
+
+        // Add authenticated data
+        cipher.updateAAD(aad);
+
+        // And try to verify & decrypt
+        byte[] plaintext;
+        try {
+            plaintext = cipher.doFinal(transferFrame, dataStart, secTrailerEnd - dataStart);
+        } catch (AEADBadTagException e) {
+            return VerificationStatusCode.MacVerificationFailure;
+        } catch (IllegalBlockSizeException | BadPaddingException e) {
+            return VerificationStatusCode.DecryptionFailed;
+        }
+
+        // Copy the decrypted plaintext back into the frame
+        System.arraycopy(plaintext, 0, transferFrame, dataStart, plaintext.length);
+
+        // Check the sequence number
+        int receivedSeqNum = ByteArrayUtils.decodeInt(transferFrame, secHeaderStart + 2 + GCM_IV_LEN_BYTES);
+        int minAllowedSeqNum = seqNum - seqNumWindow;
+        int maxAllowedSeqNum = seqNum + seqNumWindow;
+
+        if (receivedSeqNum < minAllowedSeqNum || receivedSeqNum > maxAllowedSeqNum) {
+            log.error("Received sequence number {} outside of range [{}..{}]", receivedSeqNum, minAllowedSeqNum,
```

Comment:
In Yamcs we use log.error for things which are almost certain bugs. Here it should be log.warning